### PR TITLE
Replace `crates-index` with `tame-index`

### PR DESCRIPTION
**Pull request by Jake-Shadle**
This PR replaced `crates-index` with `tame-index` which gives more friendly access to the crates.io index, whether it is git, sparse, or source replaced with another git or sparse registry.

Resolves: #487

### CODE REVIEW

**Comment by Jake-Shadle** [comment]
Oh, didn't realize this crate tested in several old versions, I'll relax it to 1.67.0.

**Comment by obi1kenobi** [comment]
> This was fixed in [EmbarkStudios@a966a4e](https://github.com/EmbarkStudios/cargo-semver-checks/commit/a966a4e840dc1aa29980aeb4787b6d6340be3233), not sure why the commit isn't being picked up in this PR...

Another contributor in another repo reported the same problem yesterday. I have to assume this is a GitHub bug.

A retry like `git commit --amend --no-edit` followed by a force-push of the branch seemed to resolve it that time.

File: src/rustdoc_gen.rs [hunk 1]
```diff
@@ -597,13 +600,56 @@ impl core::fmt::Debug for RustdocFromRegistry {
...
-        let mut index = crates_index::Index::new_cargo_default()?;
+        let index_url = tame_index::IndexUrl::crates_io(
+            // This is the config root, where .cargo/config.toml configuration files
+            // are crawled to determine if crates.io has been source replaced
+            // <https://doc.rust-lang.org/cargo/reference/source-replacement.html>
+            // if not specified it defaults to the current working directory,
+            // which is the same default that cargo uses, though note this can be
+            // extremely confusing if one can specify the manifest path of the
+            // crate from a different current working directory, though AFAICT
+            // this is not how this binary works
```
Comment (obi1kenobi) [inline src/rustdoc_gen.rs:611]: The `--manifest-path` flag can be used to select a crate from a different working directory, so I don't believe this is accurate. (The plan is to merge `cargo-semver-checks` into cargo itself, so our flags try as much as possible to behave as cargo's own.)

At the point where `RustdocFromRegistry` is constructed, the `--manifest-path` value (if any) is not present anymore. But as one possible solution, we could put the `index_url` value inside `cargo_semver_checks::Check` and then inside `GlobalConfig`, which will make it available throughout. Or if the better thing to carry around is a `ComboIndexCache`, we could do the same with that struct.

Just one thing to note: it's possible to use the binary and library here without an index. For example, if the user has already generated baseline and current rustdoc JSON files and just wants to semver-check them. This shouldn't fail with a network or configuration error on account of the (unused) index.

**Review by obi1kenobi — COMMENTED** [review]
This seems very nice! I like many of the changes you've decided to make relative to `crates-index` and I'm excited to merge this.

I flagged one concern that I believe should be possible to resolve without too much additional trouble.

Also, one more question: would this PR also enable our users to use mirrors of crates.io via source replacement? That's been a commonly requested feature (#160) and I was wondering if this PR might also happen to resolve that as well.

File: src/rustdoc_gen.rs (hunk 1 above)
Comment (Jake-Shadle) [reply src/rustdoc_gen.rs:611]: Yah I noticed that trying to debug why a test was failing, I think for now I'll keep this as it is since it will follow the same cargo logic of using configuration based on the current directory rather than the parent of the manifest, but IMO that logic is incredibly confusing if you don't know about it.

**Comment by Jake-Shadle** [comment]
> Also, one more question: would this PR also enable our users to use mirrors of crates.io via source replacement? That's been a commonly requested feature (https://github.com/obi1kenobi/cargo-semver-checks/issues/160) and I was wondering if this PR might also happen to resolve that as well.

For source replacement of crates.io, yes, this PR resolves that ([IndexUrl::crates_io](https://docs.rs/tame-index/0.2.2/tame_index/index/location/enum.IndexUrl.html#method.crates_io)), however #160 could mean both source replacement of crates.io or custom registries in addition to or instead of crates.io, which this PR alone wouldn't resolve, as each unique index would need to be opened based on the set of unique source indices, similar to what [this code](https://github.com/EmbarkStudios/cargo-deny/blob/3fce2031de75932099e2cf1bbd905f38f339fc2a/src/advisories/helpers/index.rs#L13-L68) is doing in cargo-deny.

File: src/rustdoc_gen.rs (hunk 1 above)
Comment (obi1kenobi) [reply src/rustdoc_gen.rs:611]: Indeed, I did not know about it and I am finding it confusing! But I agree, matching what cargo does is the right call here.

**Comment by obi1kenobi** [comment]
The code looks good to me! If you could try to amend your latest commit and force-push it to work around this GitHub glitch we seem to be having, I'm happy to merge as soon as CI passes.

Thanks again for letting me know about your fork of `crates-index` and for making this PR. `cargo-semver-checks` is much better off for your expertise and hard work here 🚀 

If you might be open to adding support for custom registries at any point, I'd love to have your help on that as well!

**Comment by Jake-Shadle** [comment]
Sorry should have been more clear, the 1.67.0 problem was because of rust-version is fixed, however there is a legitimate bug in the current code that my testing didn't catch because of some peculiarities with cargo, fixing that now before rerunning the tests here.

**Comment by Jake-Shadle** [comment]
Ok, so the error in https://github.com/obi1kenobi/cargo-semver-checks/actions/runs/5671264307/job/15368054854?pr=506 is resolved at the head of the branch I patched to since I could repro the CI environment locally in a container, the issue seems to be that it is using a cached binary where the cache key isn't accounting for Cargo.lock changes, so not sure how to fix that, but I'll publish the fixed version and change this PR to use that published version.

**Comment by Jake-Shadle** [comment]
Oh, and this PR broke the 1.64.0 check, I thought this crate only cared about 1.65.0 so set the rust-version to that, I'll update it to 1.64.0 as otherwise that job will succeed, but by failing on an error that bypasses the actual failure that job is intended to test.

**Comment by Jake-Shadle** [comment]
Oh wait, furthest tame-index can go back is 1.65.0, which is the lowest version supported by gix https://github.com/search?q=repo%3AByron%2Fgitoxide%20rust-version&type=code

**Comment by Jake-Shadle** [comment]
Ahh, so nvm looking how the caching works the cache key is correct and it is using the correct version, I'll continue taking a look tomorrow.

**Comment by obi1kenobi** [comment]
> Oh, and this PR broke the 1.64.0 check, I thought this crate only cared about 1.65.0 so set the rust-version to that, I'll update it to 1.64.0 as otherwise that job will succeed, but by failing on an error that bypasses the actual failure that job is intended to test.

That check was actually buggy in itself, fixed in #507 — thanks for getting me to take a look 😅 The actual min supported version is 1.67 and we don't need support for anything older than that.

`cargo-semver-checks` has two Rust versions it cares about: the version it's compiled with, and the version found at runtime which will generate the rustdoc JSON. That check ensures that if the _runtime_ rustc version is too old, we fail with a better error than "failed to deserialize rustdoc JSON." It tests a binary compiled with a newer rustc (e.g. like a prebuilt binary) running on a system with an older rustc (previously 1.64, now 1.66 in #507).

So supporting 1.67 in `tame-index` should be totally fine.

> Ahh, so nvm looking how the caching works the cache key is correct and it is using the correct version, I'll continue taking a look tomorrow.

Yeah, as you probably already saw, we just build the binary once in the workflow (per-run) and reuse it in multiple jobs to skip needing to compile from scratch every time. The failure is worth looking into, though no particular rush -- tomorrow is fine. Thanks for digging into this!

**Comment by Jake-Shadle** [comment]
Ok, I think this is ready to go now, doing this PR forced me to fix several bugs and add improvements to tame-index, eg https://github.com/EmbarkStudios/tame-index/pull/7, so it's now in a much better state.

**Comment by Jake-Shadle** [comment]
If you'd like, I can also do a follow up PR to replace git2 with gix to reduce compile times/binary size.

**Comment by obi1kenobi** [comment]
This looks great, thanks so much for digging into the issues and fixing them! I'm very excited to merge and release this 🚀 

Is my understanding correct that this PR switches our index behavior from "always git index" to "same as cargo" i.e. default-sparse unless configured otherwise?

If so, then I think all our CI tests are now over a sparse index and we have a gap in our test suite. Would you mind adding a test job that makes sure semver-checking with a git index still works as expected? For example, perhaps duplicate the `run-on-ref-slice-fork` and `run-on-ref-slice-fork-windows` jobs and run variants of them that first configure a git index?

**Comment by Jake-Shadle** [comment]
The `cross-version-caching` still uses the git index as it is using 1.67 which hadn't yet stabilized sparse indices, much less made them the default, as seen by the [failure](https://github.com/obi1kenobi/cargo-semver-checks/actions/runs/5671567612/job/15369018687) before I resolved the issue in tame-index.

**Comment by obi1kenobi** [comment]
Unfortunately, that is only temporary coverage — this next release is the last one that will support 1.67 — so that job will get updated to use two newer Rust versions that happen to emit different rustdoc JSON formats.

I'd really love an explicit test job for testing the git index, because bugs related to index configuration are always such a pain to triage and debug.

How can I make adding that test as easy as possible for you?

I've never added or modified index configs before, so I don't know how to assert that cargo / cargo-semver-checks are using a git index. But if you'd prefer, I can add two new jobs (Linux / Windows) that catch semver issues and have a fill-in-the-blank step that should add the configuration and assert it's being applied correctly.

**Comment by Jake-Shadle** [comment]
Oh it's no problem for me to add them I just thought I would point out there is still at least one test using the git index, I'll add variations for the above tests you recommended.

**Comment by obi1kenobi** [comment]
Thanks, I really appreciate it!

Also, switching to `gix` in a separate PR would be great as well!

**Comment by obi1kenobi** [comment]
Very interesting, TIL about that env var!

Is `tame-index` guaranteed to read it? Is there a way to assert that it has?

I'm a bit worried about "test passes for the wrong reasons" situations, like accidentally ignoring the env var and falling back to the default index config.

**Comment by Jake-Shadle** [comment]
Yes, the purpose of the `IndexUrl::crates_io` is to open the same index that cargo would given the same environment, including the [env var](https://github.com/EmbarkStudios/tame-index/blob/main/src/index/location.rs#L64-L69). I understand your concern though, but I think in this particular case it's not actually very important, the index information that is used is identical whether it used the git or sparse index for crates.io.

**Review by obi1kenobi — APPROVED** [review]
Thanks for all the awesome work you did. `cargo-semver-checks` and all its users are much better off for it 🚀

GitHub is showing a merge conflict at the moment, but as soon as that's resolved, I'm happy to merge this.

**Comment by obi1kenobi** [comment]
Thanks again!

If you're still interested in porting `cargo-semver-checks` to `gix`, I'd happily merge a PR for that. But obviously, no pressure at all.

**Comment by Jake-Shadle** [comment]
Next week ;)

**Comment by obi1kenobi** [comment]
Enjoy your weekend!
